File: Reliability/300_Testing_for_Resiliency_of_EC2_RDS_and_S3/Code/FailureSimulations/java/appresiliency/src/main/java/com/app/resiliency/AZFailover.java
//
// MIT No Attribution
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy of this
// software and associated documentation files (the "Software"), to deal in the Software
// without restriction, including without limitation the rights to use, copy, modify,
// merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
// permit persons to whom the Software is furnished to do so.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
// INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
// SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//

package com.app.resiliency;

import com.amazonaws.services.autoscaling.AmazonAutoScaling;
import com.amazonaws.services.autoscaling.AmazonAutoScalingClientBuilder;
import com.amazonaws.services.autoscaling.model.AutoScalingGroup;
import com.amazonaws.services.autoscaling.model.UpdateAutoScalingGroupRequest;
import com.amazonaws.services.ec2.AmazonEC2;
import com.amazonaws.services.ec2.AmazonEC2ClientBuilder;
import com.amazonaws.services.ec2.model.*;
import com.amazonaws.services.rds.AmazonRDS;
import com.amazonaws.services.rds.AmazonRDSClientBuilder;
import com.amazonaws.services.rds.model.DBInstance;
import com.amazonaws.services.rds.model.DescribeDBInstancesResult;
import com.amazonaws.services.rds.model.RebootDBInstanceRequest;
import org.apache.commons.lang3.StringUtils;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.LinkedList;
import java.util.Arrays;

public class AZFailover {

    private static final AmazonEC2 EC2_CLIENT = AmazonEC2ClientBuilder.defaultClient();
    private static final AmazonRDS RDS_CLIENT = AmazonRDSClientBuilder.defaultClient();
    private String vpcId;
    private String azId;

    AZFailover(String vpcId, String azId) {
        this.vpcId = vpcId;
        this.azId = azId;
    }

    public void failover() {

        try {
            AmazonAutoScaling autoScalingclient = AmazonAutoScalingClientBuilder.standard().build();

            // Modify the autoscaling group to remove the AZ affected which is the AZ passed in the input
            // Find the autoscaling group that this is deployed into
            if (autoScalingclient.describeAutoScalingGroups() != null
                    && autoScalingclient.describeAutoScalingGroups().getAutoScalingGroups().size() > 0) {
                // Note: This assumes a group exists for readability
                AutoScalingGroup autoScalingGroup = autoScalingclient.describeAutoScalingGroups().getAutoScalingGroups().get(0);
                String autoScalingGroupName = autoScalingGroup.getAutoScalingGroupName();

                // Find all subnets in the availability zone passed in the input
                DescribeSubnetsResult describeSubnetsResult
                        = EC2_CLIENT.describeSubnets(new DescribeSubnetsRequest().withFilters(
                        new Filter("vpc-id", Collections.singletonList(vpcId))));
                List<String> desiredSubnetsForASG = new ArrayList<>();
                for (Subnet subnet : describeSubnetsResult.getSubnets()) {
                    if (!subnet.getAvailabilityZone().equalsIgnoreCase(azId)) {
                        desiredSubnetsForASG.add(subnet.getSubnetId());
                    }
                }

                String[] vpcZoneIdentifiers = autoScalingGroup.getVPCZoneIdentifier().split(",");
                List<String> desiredSubnets = new LinkedList<>(Arrays.asList(vpcZoneIdentifiers));

                for (String subnet : desiredSubnets) {
                    if(!desiredSubnetsForASG.contains(subnet)) {
                        desiredSubnets.remove(subnet);
                    }
                }

                UpdateAutoScalingGroupRequest request = new UpdateAutoScalingGroupRequest().withAutoScalingGroupName(autoScalingGroupName)
                    .withVPCZoneIdentifier(StringUtils.join(desiredSubnets, ','));
                System.out.println("Updating the auto scaling group " + autoScalingGroupName + " to remove the subnet in the AZ");
                autoScalingclient.updateAutoScalingGroup(request);
            }

            // Find all subnets in the availability zone passed in the input
            DescribeSubnetsResult describeSubnetsResult
                    = EC2_CLIENT.describeSubnets(new DescribeSubnetsRequest().withFilters(
                    new Filter("vpc-id", Collections.singletonList(vpcId)),
                    new Filter("availabilityZone", Collections.singletonList(azId))));
            List<String> desiredSubnetsForAddingNewNacl = new ArrayList<>();
            for (Subnet subnet : describeSubnetsResult.getSubnets()) {
                desiredSubnetsForAddingNewNacl.add(subnet.getSubnetId());
            }

            //Find all the network acl associations matching the subnets identified above
            DescribeNetworkAclsResult describeNetworkAclsResult
                    = EC2_CLIENT.describeNetworkAcls(new DescribeNetworkAclsRequest().withFilters(
                    new Filter("association.subnet-id", desiredSubnetsForAddingNewNacl)));

            List<NetworkAclAssociation> desiredAclAssociations = new ArrayList<>();
            // Note: This assumes a Network ACL exists for readability
            List<NetworkAclAssociation> networkAclsAssociatedWithSubnet = describeNetworkAclsResult.getNetworkAcls().get(0).getAssociations();
            for (String subnetId : desiredSubnetsForAddingNewNacl) {
                for (NetworkAclAssociation networkAcl : networkAclsAssociatedWithSubnet) {
                    if (networkAcl.getSubnetId().equalsIgnoreCase(subnetId)) {
                        desiredAclAssociations.add(networkAcl);
                    }
                }
            }

            //create new network acl association with both ingress and egress denying to all the traffic
            CreateNetworkAclRequest createNetworkAclRequest = new CreateNetworkAclRequest();
            createNetworkAclRequest.setVpcId(vpcId);
            // Note: This assumes a Network ACL exists for readability
            String networkAclId = EC2_CLIENT.createNetworkAcl(createNetworkAclRequest).getNetworkAcl().getNetworkAclId();
            createNetworkAclEntry(networkAclId, 100, "0.0.0.0/0", true, "-1",
                    createPortRange(0, 65535), RuleAction.Deny);
            createNetworkAclEntry(networkAclId, 101, "0.0.0.0/0", false, "-1",
                    createPortRange(0, 65535), RuleAction.Deny);

            // replace all the network acl associations identified for the above subnets with the new network
            // acl association which will deny all traffic for those subnets in that AZ
            System.out.println("Creating new network ACL associations");
            replaceNetworkAclAssociations(desiredAclAssociations, networkAclId);

            //fail over rds which is in the same AZ
            DescribeDBInstancesResult describeDBInstancesResult = RDS_CLIENT.describeDBInstances();
            List<DBInstance> dbInstances = describeDBInstancesResult.getDBInstances();
            String dbInstancedId = null;
            for (DBInstance dbInstance : dbInstances) {
                if(dbInstance.getDBSubnetGroup().getVpcId().equalsIgnoreCase(vpcId)
                        && dbInstance.getAvailabilityZone().equalsIgnoreCase(azId)
                        && dbInstance.getStatusInfos().isEmpty()
                        && dbInstance.getMultiAZ().booleanValue()) {
                    dbInstancedId = dbInstance.getDBInstanceIdentifier();
                }
            }
            // we want to fail over rds if rds is present in the same az where it is affected
            if (dbInstancedId != null) {
                RebootDBInstanceRequest rebootDBInstanceRequest = new RebootDBInstanceRequest();
                rebootDBInstanceRequest.setDBInstanceIdentifier(dbInstancedId);
                rebootDBInstanceRequest.setForceFailover(true);
                System.out.println("Rebooting dbInstanceId to secondary AZ" + dbInstancedId);
                RDS_CLIENT.rebootDBInstance(rebootDBInstanceRequest);
            }
        } catch (Exception exception) {
            System.out.println("Unknown exception occurred " + exception.getMessage());
        }

    }

    private PortRange createPortRange(int from, int to) {
        PortRange portRange = new PortRange();
        portRange.setFrom(from);
        portRange.setTo(to);
        return portRange;
    }

    private void createNetworkAclEntry(String networkAclId, Integer ruleNumber, String cidrBlock, boolean egress,
            String protocol, PortRange portRange, RuleAction ruleAction) {
        CreateNetworkAclEntryRequest createNetworkAclEntryRequest = new CreateNetworkAclEntryRequest();
        createNetworkAclEntryRequest.setNetworkAclId(networkAclId);
        createNetworkAclEntryRequest.setRuleNumber(ruleNumber);
        createNetworkAclEntryRequest.setCidrBlock(cidrBlock);
        createNetworkAclEntryRequest.setEgress(egress);
        createNetworkAclEntryRequest.setProtocol(protocol);
        createNetworkAclEntryRequest.setPortRange(portRange);
        createNetworkAclEntryRequest.setRuleAction(ruleAction);
        EC2_CLIENT.createNetworkAclEntry(createNetworkAclEntryRequest);
    }

    private void replaceNetworkAclAssociations(List<NetworkAclAssociation> desiredAclAssociations, String networkAclId) {
        for (NetworkAclAssociation networkAclAssociation : desiredAclAssociations) {
            ReplaceNetworkAclAssociationRequest replaceNetworkAclAssociationRequest
                    = new ReplaceNetworkAclAssociationRequest();
            replaceNetworkAclAssociationRequest.setAssociationId(networkAclAssociation.getNetworkAclAssociationId());
            replaceNetworkAclAssociationRequest.setNetworkAclId(networkAclId);
            EC2_CLIENT.replaceNetworkAclAssociation(replaceNetworkAclAssociationRequest);
        }
    }

}

File: static/Reliability/300_Testing_for_Resiliency_of_EC2_RDS_and_S3/Code/FailureSimulations/c#/AppResiliency/AppResiliency/AZFailover.cs
//
// MIT No Attribution
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy of this
// software and associated documentation files (the "Software"), to deal in the Software
// without restriction, including without limitation the rights to use, copy, modify,
// merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
// permit persons to whom the Software is furnished to do so.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
// INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
// SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//

using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Amazon.AutoScaling;
using Amazon.AutoScaling.Model;
using Amazon.EC2;
using Amazon.EC2.Model;
using Amazon.RDS;
using Amazon.RDS.Model;

namespace com.app.resiliency
{

    public class AZFailover
    {
        private IAmazonEC2 ec2Client;
        private IAmazonRDS rdsClient;
        private IAmazonAutoScaling asClient;
        private string vpcId;
        private string azName;

        internal AZFailover(string vpcId, string azId, Amazon.RegionEndpoint region = null)
        {
            this.vpcId = vpcId;
            this.azName = azId;

            if (region == null)
            {
                region = Amazon.RegionEndpoint.USEast2;
            }

            ec2Client = new AmazonEC2Client(region);
            rdsClient = new AmazonRDSClient(region);
            asClient = new AmazonAutoScalingClient(region);
        }

        public virtual async Task Failover()
        {
            try
            {
                Console.WriteLine($"Simulating AZ failure for {azName}");

                // Find all subnets in the availability zone passed in the input
                DescribeSubnetsResponse subnetsInSpecifiedAZandVpc
                        = await ec2Client.DescribeSubnetsAsync(new DescribeSubnetsRequest()
                        {
                            Filters = new List<Amazon.EC2.Model.Filter> {
                                    new Amazon.EC2.Model.Filter {
                                        Name = "vpc-id",
                                        Values = { vpcId }
                                    },
                                    new Amazon.EC2.Model.Filter {
                                        Name = "availability-zone",
                                        Values = { azName }
                                    }
                            }
                        });

                List<string> subnetIdsInSpecifiedAZ = subnetsInSpecifiedAZandVpc.Subnets.Select(x => x.SubnetId).ToList();

                // Modify the autoscaling group to remove the AZ affected which is the AZ passed in the input
                // Find the autoscaling group that this is deployed into
                DescribeAutoScalingGroupsResponse autoScalingGroupsResponse = await asClient.DescribeAutoScalingGroupsAsync();

                if (autoScalingGroupsResponse != null && autoScalingGroupsResponse.AutoScalingGroups.Count > 0)
                {
                    // Note: This assumes an Auto Scaling group exists; no error checking for readability
                    AutoScalingGroup autoScalingGroup = autoScalingGroupsResponse.AutoScalingGroups[0];

                    Console.WriteLine($"Updating the auto scaling group {autoScalingGroup.AutoScalingGroupName} to remove the subnet in {azName}");

                    UpdateAutoScalingGroupResponse updateAutoScalingGroupResponse
                        = await asClient.UpdateAutoScalingGroupAsync(new UpdateAutoScalingGroupRequest
                        {
                            AutoScalingGroupName = autoScalingGroup.AutoScalingGroupName,
                            VPCZoneIdentifier = String.Join(",", autoScalingGroup.VPCZoneIdentifier.Split(',').Where(x => !subnetIdsInSpecifiedAZ.Contains(x)))
                        });
                }

                Console.WriteLine("Creating new network ACL associations");
                await BlockSubnetsInAZ(vpcId, subnetIdsInSpecifiedAZ);

                Console.WriteLine("Failing over database");

                //fail over rds which is in the same AZ
                DescribeDBInstancesResponse describeDBInstancesResult = await rdsClient.DescribeDBInstancesAsync();

                string dbInstancedId = describeDBInstancesResult.DBInstances.Where(x => String.Equals(x.DBSubnetGroup.VpcId, vpcId, StringComparison.OrdinalIgnoreCase) &&
                    String.Equals(x.AvailabilityZone, azName, StringComparison.OrdinalIgnoreCase) &&
                    x.MultiAZ &&
                    !x.StatusInfos.Any())?.Select(x => x.DBInstanceIdentifier).FirstOrDefault();

                // we want to fail over rds if rds is present in the same az where it is affected
                if (!String.IsNullOrEmpty(dbInstancedId))
                {
                    Console.WriteLine("Rebooting dbInstanceId to secondary AZ " + dbInstancedId);

                    var response = await rdsClient.RebootDBInstanceAsync(new RebootDBInstanceRequest()
                    {
                        DBInstanceIdentifier = dbInstancedId,
                        ForceFailover = true
                    });
                }
                else
                {
                    Console.WriteLine($"Didn't find DB in the same AZ as {azName}");
                }

                Console.Write("Done");
            }
            catch (Exception exception)
            {
                Console.WriteLine("Unkown exception occured " + exception.Message);
            }
        }

        private async Task BlockSubnetsInAZ(string vpcId, List<string> subnetIds)
        {
            //Find all existing network acl associations matching the subnets identified above
            DescribeNetworkAclsResponse describeNetworkAclsResult
                = await ec2Client.DescribeNetworkAclsAsync(new DescribeNetworkAclsRequest()
                {
                    Filters = new List<Amazon.EC2.Model.Filter> {
                                    new Amazon.EC2.Model.Filter {
                                        Name = "association.subnet-id",
                                        Values = subnetIds
                                    }
                            }
                });

            // The describe will return all associations of an ACL, which can be associated with a subnet not in the filter
            IEnumerable<string> associationsToUpdate = describeNetworkAclsResult.NetworkAcls.SelectMany(x => x.Associations).Where(x => subnetIds.Contains(x.SubnetId)).Select(x => x.NetworkAclAssociationId);

            //create new network acl 
            CreateNetworkAclResponse createNetworkAclResponse = await ec2Client.CreateNetworkAclAsync(new CreateNetworkAclRequest()
            {
                VpcId = vpcId
            });

            // add both ingress and egress denying to all the traffic to the new ACL
            string networkAclId = createNetworkAclResponse.NetworkAcl.NetworkAclId;
            await CreateNetworkAclEntry(networkAclId, 100, "0.0.0.0/0", true, "-1", CreatePortRange(0, 65535), RuleAction.Deny);
            await CreateNetworkAclEntry(networkAclId, 101, "0.0.0.0/0", false, "-1", CreatePortRange(0, 65535), RuleAction.Deny);

            // update all subnets to be associated with the new ACL
            foreach (string existingAssociation in associationsToUpdate)
            {
                // associates the specified network ACL with the subnet for the specified network ACL association
                ReplaceNetworkAclAssociationResponse replaceNetworkAclAssociationResponse
                    = await ec2Client.ReplaceNetworkAclAssociationAsync(new ReplaceNetworkAclAssociationRequest()
                    {
                        AssociationId = existingAssociation,
                        NetworkAclId = networkAclId
                    });
            }
        }

        private PortRange CreatePortRange(int from, int to)
        {
            PortRange portRange = new PortRange();
            portRange.From = from;
            portRange.To = to;
            return portRange;
        }

        private async Task CreateNetworkAclEntry(string networkAclId, int ruleNumber, string cidrBlock, bool egress, string protocol, PortRange portRange, RuleAction ruleAction)
        {
            CreateNetworkAclEntryRequest createNetworkAclEntryRequest = new CreateNetworkAclEntryRequest();
            createNetworkAclEntryRequest.NetworkAclId = networkAclId;
            createNetworkAclEntryRequest.RuleNumber = ruleNumber;
            createNetworkAclEntryRequest.CidrBlock = cidrBlock;
            createNetworkAclEntryRequest.Egress = egress;
            createNetworkAclEntryRequest.Protocol = protocol;
            createNetworkAclEntryRequest.PortRange = portRange;
            createNetworkAclEntryRequest.RuleAction = ruleAction;
            CreateNetworkAclEntryResponse createNetworkAclEntryResponse = await ec2Client.CreateNetworkAclEntryAsync(createNetworkAclEntryRequest);
        }
    }
}
